File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/1360-maximum-length-of-a-concatenated-string-with-unique-characters.java
class Solution {
    int len = 0;
    int[] chars = new int[26];
    int longest = 0;
    HashSet<Character> set = new HashSet<>();
    public int maxLength(List<String> arr) {
        backtrack(arr, 0);
        return longest;
    }

    private void backtrack(List<String> arr, int idx) {
        if (len > longest) {
            longest = len;
        }
        if (idx == arr.size()) {
            
            return;
        }

        for (int i = idx; i < arr.size(); i++) {
            String s = arr.get(i);
            if (!contains(s) && !duplicate(s)) {
                for (int j = 0; j < s.length(); j++) {
                    chars[s.charAt(j) - 'a']++;
                }
                len += s.length();
                backtrack(arr, i + 1);
                len -= s.length();
                for (int j = 0; j < s.length(); j++) {
                    chars[s.charAt(j) - 'a']--;
                }
            }
        }
    }

    private boolean contains(String s) {
        for (int i = 0; i < s.length(); i++) {
            if (chars[s.charAt(i) - 'a'] > 0) {
                return true;
            }
        }

        return false;
    }

    private boolean duplicate(String s) {
         char[] str = s.toCharArray();
         for(char c : str) {
             if (!set.add(c)) {
                 set.clear();
                 return true;
             }
         }

        set.clear();
        return false;

    }
}

File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/1360-maximum-length-of-a-concatenated-string-with-unique-characters.py
class Solution:
    # def __init__(self):
    #     self.maxLen = 0
    #     self.currLen = 0
    #     self.used = [0] * 26

    # def maxLength(self, arr: List[str]) -> int:
    #     self.backtracking(arr, 0)
    #     return self.maxLen
    
    # def backtracking(self, arr: List[str], idx: int):
    #     self.maxLen = max(self.maxLen, self.currLen)

    #     if idx == len(arr):
    #         return
        
    #     for i in range(idx, len(arr)):
    #         if self.isValid(arr[i]):
    #             self.currLen += len(arr[i])
    #             for j in range(len(arr[i])):
    #                 self.used[ord(arr[i][j]) - ord('a')] = 1
    #             self.backtracking(arr, i + 1)
    #             self.currLen -= len(arr[i])
    #             for j in range(len(arr[i])):
    #                 self.used[ord(arr[i][j]) - ord('a')] = 0
    
    # def isValid(self, s: str) -> bool:
    #     counts = [0] * 26
    #     for i in range(len(s)):
    #         if self.used[ord(s[i]) - ord('a')] > 0 or counts[ord(s[i]) - ord('a')]:
    #             return False
    #         counts[ord(s[i]) - ord('a')] += 1

    #     return True
    def maxLength(self, arr: List[str]) -> int:
        # Convert strings to bitmask representation
        masks = []
        for s in arr:
            mask = 0
            for ch in s:
                # Check if character is seen before in the same string
                if (mask >> (ord(ch) - ord('a'))) & 1:
                    mask = 0
                    break
                mask |= 1 << (ord(ch) - ord('a'))
            if mask:
                masks.append(mask)

        # Backtracking
        def backtrack(idx, current_mask):
            if idx == len(masks):
                return 0
            # Without current string
            ans = backtrack(idx + 1, current_mask)
            # With current string, if no overlap of characters
            if not (current_mask & masks[idx]):
                ans = max(ans, bin(masks[idx]).count('1') + backtrack(idx + 1, current_mask | masks[idx]))
            return ans

        return backtrack(0, 0)